_posts/2025-05-05-plan-execute-agent.md

```diff
@@ -0,0 +1,447 @@
+
+* Break down a complex investigation into logical steps
+* Automatically select and use appropriate tools
+* Analyze multiple data sources (logs, traces, metrics)
+* Provide clear, actionable insights
+
+Whether you're managing cloud infrastructure, developing applications, or building AI-powered solutions, this new capability can help streamline your workflows and enhance your problem-solving capabilities.
+
+Now let's explore the capabilities and workflow of the plan-execute-reflect agent.
+
+## What is the plan-execute-reflect agent?
+
+The [plan-execute-reflect](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/agents/plan-execute-reflect/) agent is a long-running agent designed for complex, multi-step tasks. This agent is capable of breaking down a complex task into a series of simple steps (plan), executing each step, and optimizing its plan based on intermediate step results. It uses a separate executor agent for the execution of substeps.
+
+Key features of this agent include the following:
+
+* Adaptive planning that evolves based on intermediate results
+* Flexibility to use different models for planning and execution phases by specifying your own executor agent during registration (uses a [conversational](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/agents/conversational/) agent by default)
+* Asynchronous execution capability for handling long-running workflows in the background (when `async=true`)
+* Built-in MCP client functionality allowing connections to multiple MCP servers
+* Standardized tool communication through function calling
+* Support for custom prompts, enabling specialization for specific use cases such as Observability Agent or Research Agent (using WebSearchTool)
+
+The following diagram illustrates the execution flow of the plan-execute-reflect agent.
+
+![PlanExecuteReflect Agent Execution Workflow](/assets/media/blog-images/2025-05-05-plan-execute-agent/per-agent-execution-workflow.png){: .img-fluid}
+
+The workflow consists of the following steps:
+
+1. A user provides the plan-execute-reflect agent with a task.
+2. The plan-execute-reflect agent forwards the task to the planner LLM.
+3. The LLM returns a plan (a series of steps to execute).
+4. The plan-execute-reflect agent forwards the first step of the plan to the executor agent.
+5. The executor agent executes the steps and returns the response.
+6. The plan-execute-reflect agent forwards the result of the executed step and the original plan to the planner LLM.
+7. The planner LLM either returns the final result or a refined plan.
+8. If the planner LLM returns the final result, the results is returned to the user. Otherwise, the planner LLM returns a new plan and executes the steps (steps 4--7 on the diagram) until it has enough information to complete the task and return the result.
+
```

Comment:
Line 70, 2nd sentence: "returns a new plan and executes steps 4--7 until it has enough information to complete the task and return the result"?

Reply:
I initially wanted to keep it more general, but since we're describing the picture, I'll change to the suggested text.